rust/rust.cc

```diff
@@ -0,0 +1,323 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2022  Lofty <lofty@yosyshq.com>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "log.h"
+#include "nextpnr.h"
+
+namespace {
+    USING_NEXTPNR_NAMESPACE;
+
+    template<typename T>
+    uint64_t wrap(T thing) {
+        static_assert(sizeof(T) <= 8, "T is too big for FFI");
+        uint64_t b = 0;
+        memcpy(&b, &thing, sizeof(T));
```

Comment:
I'd be curious if you could check whether a `reinterpret_cast` type approach works out faster than `memcpy` or not? I'd expect the compiler can optimise this away anyway though? Might be worth a check in godbolt or whatever.

Reply:
The compiler can absolutely optimise this away, and `memcpy` is pretty much what the standard requires for strict aliasing. Not that the rest of this code is necessarily standard-kosher.

Reply:
Thanks for confirming!